### PR TITLE
ISPN-7172 Total order caches can hang during join

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-7172

* Fix WithinThreadExecutor handling in LimitedExecutor
* Remove LimitedExecutor permit before putting the LocalCacheStatus
  in the runningCaches map.

### CODE REVIEW

File: core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java [hunk 1]
```diff
@@ -130,15 +130,17 @@ public CacheTopology join(String cacheName, CacheJoinInfo joinInfo, CacheTopolog
...
 
-      long timeout = joinInfo.getTimeout();
-      long endTime = timeService.expectedEndTime(timeout, TimeUnit.MILLISECONDS);
       // Pretend the join is using up a thread from the topology updates executor.
       // This ensures that the initial topology and the GET_CACHE_LISTENERS request will happen on this thread,
-      // and other topology updates are only handled after we call backgroundTaskFinished(null)
+      // and other topology updates are only handled after we complete joinFuture.
       CompletableFuture<Void> joinFuture = new CompletableFuture<>();
       cacheStatus.getTopologyUpdatesExecutor().executeAsync(() -> joinFuture);
+
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: is this change really needed?

Comment (danberindei) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: I'm not sure if it's strictly necessary for correctness at this point or it just makes tests more predictable, but I think it's safer to implement it as promised in the comments above.

Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: I don't see how initializing the  ```runningCaches``` before or after the sending the ```joinFuture``` to the executor could affect anything... 
My point is: I'm not seeing any difference in the logic executed. So, what side effect am I missing?

Comment (danberindei) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: Submitting `joinFuture` will prevent the executor from running any other task until `joinFuture` is completed. If the cache exists in `runningCaches` and it's `LimitedExecutor` has a free spot, it will process topology updates, and those have a chance of doing "stuff" before we have properly joined.
I think the initial stuff I was worried about was just blocking for a new view, which could overwhelm the OOB thread pool given enough caches.

Comment (pruivo) [inline core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java]: makes sense.

**Comment by pruivo** [comment]
integrated! thanks @danberindei !
